src/objective-c/GRPCClient/GRPCCallOptions.m

```diff
@@ -0,0 +1,431 @@
+                     authTokenProvider:kDefaultAuthTokenProvider
+                       initialMetadata:kDefaultInitialMetadata
+                       userAgentPrefix:kDefaultUserAgentPrefix
+                     responseSizeLimit:kDefaultResponseSizeLimit
+                     compressAlgorithm:kDefaultCompressAlgorithm
+                           enableRetry:kDefaultEnableRetry
+                     keepaliveInterval:kDefaultKeepaliveInterval
+                      keepaliveTimeout:kDefaultKeepaliveTimeout
+                     connectMinTimeout:kDefaultConnectMinTimeout
+                 connectInitialBackoff:kDefaultConnectInitialBackoff
+                     connectMaxBackoff:kDefaultConnectMaxBackoff
+                 additionalChannelArgs:kDefaultAdditionalChannelArgs
+                           pemRootCert:kDefaultPemRootCert
+                         pemPrivateKey:kDefaultPemPrivateKey
+                          pemCertChain:kDefaultPemCertChain
+                         transportType:kDefaultTransportType
+                      hostNameOverride:kDefaultHostNameOverride
+                            logContext:kDefaultLogContext
+                     channelPoolDomain:kDefaultChannelPoolDomain
+                             channelId:kDefaultChannelId];
+}
+
+- (instancetype)initWithServerAuthority:(NSString *)serverAuthority
+                                timeout:(NSTimeInterval)timeout
+                      oauth2AccessToken:(NSString *)oauth2AccessToken
+                      authTokenProvider:(id<GRPCAuthorizationProtocol>)authTokenProvider
+                        initialMetadata:(NSDictionary *)initialMetadata
+                        userAgentPrefix:(NSString *)userAgentPrefix
+                      responseSizeLimit:(NSUInteger)responseSizeLimit
+                      compressAlgorithm:(GRPCCompressAlgorithm)compressAlgorithm
+                            enableRetry:(BOOL)enableRetry
+                      keepaliveInterval:(NSTimeInterval)keepaliveInterval
+                       keepaliveTimeout:(NSTimeInterval)keepaliveTimeout
+                      connectMinTimeout:(NSTimeInterval)connectMinTimeout
+                  connectInitialBackoff:(NSTimeInterval)connectInitialBackoff
+                      connectMaxBackoff:(NSTimeInterval)connectMaxBackoff
+                  additionalChannelArgs:(NSDictionary *)additionalChannelArgs
+                            pemRootCert:(NSString *)pemRootCert
+                          pemPrivateKey:(NSString *)pemPrivateKey
+                           pemCertChain:(NSString *)pemCertChain
+                          transportType:(GRPCTransportType)transportType
+                       hostNameOverride:(NSString *)hostNameOverride
+                             logContext:(id)logContext
+                      channelPoolDomain:(NSString *)channelPoolDomain
+                              channelId:(NSUInteger)channelId {
+  if ((self = [super init])) {
```

Comment:
you need to copy all the arguments here that support NSCopying (all the strings). For dictionaries also consider doing the deeper copy afforded by [[NSDictionary alloc] initWithDictionary:... copyItems:YES];